cmake/EnableSanitizers.cmake

```diff
@@ -0,0 +1,86 @@
+#   Copyright 2017 Google Inc.
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+
+include(CheckCXXCompilerFlag)
+
+function(sanitizer_test NAME)
+    set(${NAME}_ENABLED FALSE PARENT_SCOPE)
+    foreach (FLAG ${ARGN})
+        set(CMAKE_REQUIRED_QUIET TRUE)
+        set(CMAKE_REQUIRED_FLAGS ${FLAG})
+        set(FNAME ${NAME}_${CMAKE_CXX_COMPILER_ID}_CXX)
+        unset(FNAME CACHE)
+        set(OLD_CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS}")
+        set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} ${FLAG}")
+        check_cxx_compiler_flag(${FLAG} ${FNAME})
+        set(CMAKE_REQUIRED_FLAGS "${OLD_CMAKE_REQUIRED_FLAGS}")
+        if (${FNAME})
+            set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${FLAG}" PARENT_SCOPE)
+            set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${FLAG}" PARENT_SCOPE)
+            set(${NAME}_ENABLED TRUE PARENT_SCOPE)
+            return()
+        endif ()
+    endforeach ()
+endfunction()
+
+option(SANITIZE_ADDRESS "Enable AddressSanitizer for the build." OFF)
+if (SANITIZE_ADDRESS)
+    set(ASAN_CANDIDATES "-fsanitize=address -fno-omit-frame-pointer" "-fsanitize=address")
+    sanitizer_test(AddressSanitizer ${ASAN_CANDIDATES})
+endif (SANITIZE_ADDRESS)
+
+option(SANITIZE_LEAKS "Enable LeakSanitizer for the build." OFF)
+if (SANITIZE_LEAKS)
+    set(ASAN_CANDIDATES "-fsanitize=leaks -fno-omit-frame-pointer" "-fsanitize=leaks")
+    sanitizer_test(LeaksSanitizer ${ASAN_CANDIDATES})
+endif (SANITIZE_LEAKS)
+
+option(SANITIZE_MEMORY "Enable MemorySanitizer for the build." OFF)
+if (SANITIZE_MEMORY)
+    if (${AddressSanitizer_ENABLED})
```

Comment:
Why is this `if (${AddressSanitizer_ENABLED})` rather than `if (SANITIZE_ADDRESS)`?

Reply:
Good question.  When I was doing this, because `SANITIZE_ADDRESS` is a request to enable it, `AddressSanitizer_ENABLED` means that it actually was (i.e., it was requested and the tests to set it passed).  Not much of a reason, I know.